[NFC] [Serializer] Pack information in serializer

ChuanqiXu9 opened the pull request:
Previously, the boolean values will occupy spaces that can contain integers. It wastes the spaces especially if the boolean values are serialized consecutively. The patch tries to pack such consecutive boolean values (and enum values) so that we can save more spaces and so the times.

Before the patch, we need 4.478s (in my machine) to build the std module (https://libcxx.llvm.org/Modules.html) with 28712 bytes for size of the BMI. After the patch, the time becomes to 4.374s and the size becomes to 27388 bytes for the size of the BMI.

This is intended to be a NFC patch.

This patch doesn't optimize all such cases. We can do it later after we have consensus on this.

ChuanqiXu9 commented:
@zygoloid @vgvassilev ping~

Personally I feel like the change itself should be safe. (The test for modules is really sensitive about the changes in Serializer)

vgvassilev commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -617,24 +617,27 @@ void ASTDeclReader::VisitDecl(Decl *D) {
-  D->InvalidDecl = Record.readInt();
-  if (Record.readInt()) { // hasAttrs
+
+  uint64_t DeclBits = Record.readInt();
+  D->InvalidDecl = DeclBits & 0x1;
+  D->setImplicit(DeclBits & (1 << 2));
+  D->Used = (DeclBits >> 3) & 0x1;
+  IsDeclMarkedUsed |= D->Used;
+  D->setReferenced(DeclBits & (1 << 4));
+  D->setTopLevelDeclInObjCContainer(DeclBits & (1 << 5));
+  D->setAccess((AccessSpecifier)((DeclBits >> 6) & 0x3));
+  D->FromASTFile = true;
+  auto ModuleOwnership = (Decl::ModuleOwnershipKind)((DeclBits >> 8) & 0x7);
```
Is there a way to improve the API avoiding these magic constants? I suspect this would be difficult to read and maintain.

ChuanqiXu9 replied:
What I had in mind is to introduce a BitPacker, but it can't avoid magic constants completely. e.g, we have to know that the width of `AccessSpecifier ` should be 2 and the width of `ModuleOwnershipKind` should be 3.

Then we can refactor the above section to:

```
BitPacker Bits = Record.readInt();
D->InvalidDecl = Bits.getNextBit();
D->setImplicit(Bits.getNextBit());
D->Used = Bits.getNextBit();
IsDeclMarkedUsed |= D->Used;
D->setReferenced(Bits.getNextBit());
D->setTopLevelDeclInObjCContainer(Bits.getNextBit());
D->setAccess((AccessSpecifier)Bits.getNextBit</*width = */2>()); // Still we can't avoid the magic constants completely.
D->FromASTFile = true;
auto ModuleOwnership = (Decl::ModuleOwnershipKind)Bits.getNextBit</*width = */3>());
```

Would you be happy about the above form?

vgvassilev replied:
That looks more readable to me. In the case of `ModuleOwnershipKind` can we use the bit size either with `sizeof` or via `NextInContextAndBits`? If we use `sizeof` this could look like:
`auto ModuleOwnership = Bits.getNextBit<Decl::ModuleOwnershipKind>());` but we will likely read/write 8 bits instead of 3 in this case...

ChuanqiXu9 replied:
Yes... and I feel the magic numbers for some types (e.g., `AccessSpecifier` and `ModuleOwnershipKind`) might not be so scaring. Since the devs can see `AccessSpecifier` and `ModuleOwnershipKind` here then they can get the size quickly. Also even if now we have to look at both ASTWriterDecl.cpp and ASTReaderDecl.cpp to change anything. I mean, it is already tightly bounded.

ChuanqiXu9 replied:
Done

vgvassilev commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2407,6 +2407,53 @@ class ASTReader
+  ~BitsUnpacker() {
+#ifndef NDEBUG
+    while (isValid())
+      assert(!getNextBit() && "There are bits failed to read!");
```
```suggestion
      assert(!getNextBit() && "There are unprocessed bits!");
```

Is that the analog of `hasUnconsumedValues`? If so, we could probably mirror it here, too.

ChuanqiXu9 replied:
Done

vgvassilev commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2407,6 +2407,53 @@ class ASTReader
+    return Ret;
+  }
+
+  bool isValidToGetNextNBits(uint32_t Width) const {
```
```suggestion
  bool canGetNextNBits(uint32_t Width) const {
```

ChuanqiXu9 replied:
Done

vgvassilev commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -1689,9 +1709,10 @@ void ASTDeclReader::VisitParmVarDecl(ParmVarDecl *PD) {
-  PD->ParmVarDeclBits.HasInheritedDefaultArg = Record.readInt();
-  if (Record.readInt()) // hasUninstantiatedDefaultArg.
+  PD->ParmVarDeclBits.IsKNRPromoted = ParmVarDeclBits.getNextBit();
+  ;
```
```suggestion
```

ChuanqiXu9 replied:
Done

vgvassilev reviewed (commented):
This looks pretty good modulo the comments.

@zygoloid ping.

vgvassilev commented on clang/lib/Serialization/ASTWriterDecl.cpp:
```diff
@@ -2005,14 +2039,7 @@ void ASTWriter::WriteDeclAbbrevs() {
-  Abv->Add(BitCodeAbbrevOp(0));                   // TopLevelDeclInObjCContainer
-  Abv->Add(BitCodeAbbrevOp(BitCodeAbbrevOp::Fixed, 2));  // AccessSpecifier
-  Abv->Add(BitCodeAbbrevOp(BitCodeAbbrevOp::Fixed, 3));  // ModuleOwnershipKind
+  Abv->Add(BitCodeAbbrevOp(BitCodeAbbrevOp::Fixed, 11)); // Decl Bits
```
These sections became a little harder to read. Perhaps we could keep the comments and this way the reader could probably do the math why this is 11 bits for example.

ChuanqiXu9 replied:
Done

ChuanqiXu9 commented on clang/lib/Serialization/ASTWriterDecl.cpp:
```diff
@@ -2005,14 +2039,11 @@ void ASTWriter::WriteDeclAbbrevs() {
-  Abv->Add(BitCodeAbbrevOp(BitCodeAbbrevOp::Fixed, 3));  // ModuleOwnershipKind
+  Abv->Add(BitCodeAbbrevOp(
+      BitCodeAbbrevOp::Fixed,
+      11)); // Packed DeclBits: isInvalidDecl, HasAttrs, isImplicit, isUsed,
```
This looks odd at the first sight. But it is formatted by clang-format. And it looks not bad after looking into it.

ChuanqiXu9 commented:
I'd like to land this in the next week if no objection come in. Since this may be safe.

vgvassilev reviewed (approved):
LGTM, thank you!

ChuanqiXu9 committed:
[NFC] [Serializer] Pack information in serializer

Previously, the boolean values will occupy spaces that can contain
integers. It wastes the spaces especially if the boolean values are
serialized consecutively. The patch tries to pack such consecutive
boolean values (and enum values) so that we can save more spaces and so
the times.

Before the patch, we need 4.478s (in my machine) to build the std module
(https://libcxx.llvm.org/Modules.html) with 28712 bytes for size of the
BMI. After the patch, the time becomes to 4.374s and the size becomes to
27388 bytes for the size of the BMI.

This is intended to be a NFC patch.

This patch doesn't optimize all such cases. We can do it later after we
have consensus on this.

ChuanqiXu9 commented:
Thanks for reviewing : )
